zipline/data/_equities.pyx

```diff
@@ -170,22 +176,42 @@ cpdef _read_bcolz_data(ctable_t table,
         intp_t first_row
         intp_t last_row
         intp_t offset
+        intp_t out_start
+        intp_t out_end
         list results = []
 
+    ndays = shape[0]
     nassets = shape[1]
     if not nassets== len(first_rows) == len(last_rows) == len(offsets):
         raise ValueError("Incompatible index arrays.")
 
     for column_name in columns:
-        raw_data = table[column_name][:]
         outbuf = zeros(shape=shape, dtype=uint32)
-        for asset in range(nassets):
-            first_row = first_rows[asset]
-            last_row = last_rows[asset]
-            offset = offsets[asset]
-            for out_idx, raw_idx in enumerate(range(first_row, last_row + 1)):
-                outbuf[out_idx + offset, asset] = raw_data[raw_idx]
-
+        if read_all:
+            raw_data = table[column_name][:]
+
+            for asset in range(nassets):
+                first_row = first_rows[asset]
+                last_row = last_rows[asset]
+                offset = offsets[asset]
+                for out_idx, raw_idx in enumerate(
+                        range(first_row, last_row + 1)):
+                    outbuf[out_idx + offset, asset] = raw_data[raw_idx]
+        else:
+            carray = table[column_name]
+
+            for asset in range(nassets):
+                first_row = first_rows[asset]
+                last_row = last_rows[asset]
+                offset = offsets[asset]
+                out_start = offset
+                out_end = (last_row - first_row) + offset + 1
```

Comment:
Also, a note the `carray` extension object does not support the buffer interface, so we still need to branch on `read_all` to get the correct types, instead of sharing an `uint32_t[:]` definition.